arrows/vxl/morphology.cxx

```diff
@@ -0,0 +1,325 @@
+                { "union", COMBINE_union },
+                { "intersection", COMBINE_intersection } );
+
+inline bool
+union_functor( bool x1, bool x2 )
+{
+  return x1 || x2;
+}
+
+inline bool
+intersection_functor( bool x1, bool x2 )
+{
+  return x1 && x2;
+}
+
+} // namespace
+
+// ----------------------------------------------------------------------------
+class morphology::priv
+{
+public:
+  priv()
+  {
+  }
+
+  ~priv();
+  // Setup structuring elements
+  void setup_internals();
+  // Compute the morpholigical operation on a single plane of an input image
+  void
+  apply_morphology_to_plane( vil_image_view< bool > const& input,
+                             vil_image_view< bool >& output,
+                             unsigned plane_index );
+  // Perform a morphological operation and optionally combine across channels
+  vil_image_view< bool >
+  perform_morphological_operations( vil_image_view< bool > const& input );
+
+  bool setup{ false };
+  vil_structuring_element morphological_element;
+
+  morphology_mode morphology_type{ MORPHOLOGY_dilate };
+  element_mode element_type{ ELEMENT_disk };
+  combine_mode combine_type{ COMBINE_none };
+  double element_size{ 1.0 };
+};
+
+// ----------------------------------------------------------------------------
+morphology::priv
+::~priv()
+{
+}
+
+// ----------------------------------------------------------------------------
+void
+morphology::priv
+::setup_internals()
+{
+  if( !setup )
+  {
+    switch( element_type )
+    {
+      case ELEMENT_disk:
+      {
+        morphological_element.set_to_disk( element_size );
+      }
+      case ELEMENT_iline:
+      {
+        morphological_element.set_to_line_i( 0, element_size );
+      }
+      case ELEMENT_jline:
+      {
+        morphological_element.set_to_line_j( 0, element_size );
+      }
+    }
+
+    setup = true;
+  }
+}
+
+// ----------------------------------------------------------------------------
+void
+morphology::priv
+::apply_morphology_to_plane( vil_image_view< bool > const& input,
+                             vil_image_view< bool >& output,
+                             unsigned plane_index )
+{
+  auto input_plane = vil_plane( input, plane_index );
+  auto output_plane = vil_plane( output, plane_index );
+  switch( morphology_type )
+  {
+    case MORPHOLOGY_erode:
+    {
+      vil_binary_erode( input_plane, output_plane, morphological_element );
+      break;
+    }
+    case MORPHOLOGY_dilate:
+    {
+      vil_binary_dilate( input_plane, output_plane, morphological_element );
+      break;
+    }
+    case MORPHOLOGY_close:
+    {
+      vil_binary_closing( input_plane, output_plane, morphological_element );
+      break;
+    }
+    case MORPHOLOGY_open:
+    {
+      vil_binary_opening( input_plane, output_plane, morphological_element );
+      break;
+    }
+    case MORPHOLOGY_none:
+    {
+      output_plane.deep_copy( input_plane );
```

Comment:
This could be refactored so that the 'none' case can just copy the whole image, which is probably more efficient. I pushed those changes to a copy of your branch in my fork; PTAL.